### PR TITLE
Implement (optional) machine-type dependent resource reservations

### PR DESCRIPTION
**Pull request by MichaelEischer**
**How to categorize this PR?**

/area robustness
/kind enhancement

**What this PR does / why we need it**:

The PR implements the "Configure sensible default values for kube-reserved based on the worker pool's machine size, similar to GKE's [formula for resource reservation](https://cloud.google.com/kubernetes-engine/docs/concepts/plan-node-sizes#resource_reservations)" part of https://github.com/gardener/gardener/issues/2590

Different from the sketched approach there, the admission plugin is enabled by default, but is configured to maintain the current behavior. The reason for that is that the defaulting in `pkg/apis/core/v1beta1/defaults_shoot.go` had to be removed, such that we can either add two admission plugins (one for the current behavior and one for using the GKE formula) or use one admission plugin but make it configurable. The latter option is probably the clearer one.

**Which issue(s) this PR fixes**:
Implements part of https://github.com/gardener/gardener/issues/2590

**Special notes for your reviewer**:

Open questions:
- [x] There's no documentation so far, should we add it as part of this PR or a follow-up one?
- [x] What do you think about the configuration option name `typeDependentReservations` of the admission plugin ? A name like `useGKEFormula` might be more descriptive.

The GKE formula in the admission controller can be enabled using the following config map (assuming that the configuration is passed to the gardener-apiserver):

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: gardener-apiserver-admission-config
  namespace: garden
data:
  configuration.yaml: |-
    apiVersion: apiserver.config.k8s.io/v1
    kind: AdmissionConfiguration
    plugins:
    - configuration:
        apiVersion: shootresourcereservation.admission.gardener.cloud/v1alpha1
        kind: Configuration
        typeDependentReservations: true
      name: ShootResourceReservation
```

**Release note**:

```feature operator
A new admission plugin `ShootResourceReservation` has been added to `gardener-apiserver`. It supports calculating resource reservations (memory/CPU/PID) for the `kubelet.kubeReserved` fields in `Shoot`s based on the available resources of a machine type. This only applies when `useGKEFormula` is set to `true`. Otherwise, the old static values remain to be used. 
```

### CODE REVIEW

File: plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Configuration provides configuration for the ShootResourceReservation admission controller.
+type Configuration struct {
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go:24]: Add an example to https://github.com/gardener/gardener/blob/master/example/20-admissionconfig.yaml

File: plugin/pkg/shoot/resourcereservation/admission.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+		return nil
+	}
+
+	if !c.typeDependentReservations {
+		setStaticResourceReservationDefaults(shoot)
+		return nil
+	}
+
+	if shoot.Spec.Kubernetes.Kubelet != nil && shoot.Spec.Kubernetes.Kubelet.KubeReserved != nil {
+		// Inject static defaults for shoots with shoot-level resource reservations
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/admission.go]:
```suggestion
		// Inject static defaults for shoots with global resource reservations
```

File: plugin/pkg/shoot/resourcereservation/admission.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+	cloudProfile, err := c.cloudProfileLister.Get(shoot.Spec.CloudProfileName)
+	if err != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not find referenced cloud profile: %+v", err.Error()))
+	}
+
+	allErrs := field.ErrorList{}
+	workersPath := field.NewPath("spec", "provider", "workers")
+
+	for i := 0; i < len(shoot.Spec.Provider.Workers); i++ {
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/admission.go:158]:
```suggestion
	for i, worker := range shoot.Spec.Provider.Workers {
```

File: plugin/pkg/shoot/resourcereservation/admission.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		if kubeReserved.PID == nil {
+			kubeReserved.PID = reservation.PID
+		}
+	}
+	return allErrs
+}
+
+func calculateResourceReservationForMachineType(cloudProfile *gardencorev1beta1.CloudProfile,
+	machineType string) (*core.KubeletConfigReserved, error) {
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/admission.go]: Put this in one line or put all arguments/return values in a single line

File: plugin/pkg/shoot/resourcereservation/admission.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+			cpuReserved := calculateCPUReservation(machine.CPU.MilliValue())
+			memoryReserved := calculateMemoryReservation(machine.Memory.Value())
+
+			return &core.KubeletConfigReserved{
+				CPU:    resource.NewMilliQuantity(cpuReserved, resource.BinarySI),
+				Memory: resource.NewQuantity(memoryReserved, resource.BinarySI),
+				PID:    &kubeReservedPID,
+			}, nil
+		}
+	}
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/admission.go]: Instead of iterating through the potentially long list of machine types again and again for each worker pool and shoot, what about building a `map[string]core.MachineType` upfront so that we can access the machine type data based on the name in `O(1)`?

File: plugin/pkg/shoot/resourcereservation/admission_test.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+)
+
+var _ = Describe("resourcereservation", func() {
+
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/admission_test.go]: remove empty line

**Review by rfranzke — CHANGES_REQUESTED** [review]
> There's no documentation so far, should we add it as part of this PR or a follow-up one?

Yes, please amend https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver_admission_plugins.md

> What do you think about the configuration option name typeDependentReservations of the admission plugin ? A name like useGKEFormula might be more descriptive.

Yes, I'd vote for explicit naming, your proposal `useGKEFormula` sounds good to me.

File: plugin/pkg/shoot/resourcereservation/admission.go (hunk 3 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/admission.go:158]: That does not work as `Workers` has type `[]Worker` such that `worker` would be a copy and not a reference to the original value in the slice. Then `injectResourceReservations` would only modify that copy.

File: plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go (hunk 1 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go:24]: I've added an example that sets `useGKEFormula` to false (which is the default anyways).

File: plugin/pkg/shoot/resourcereservation/admission.go (hunk 5 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/admission.go]: I've added a function `buildMachineTypeMap(cloudProfile *gardencorev1beta1.CloudProfile) map[string]gardencorev1beta1.MachineType`. Currently, that function runs every time the admission hook is called. Or did you have in mind to cache the map? (I'm not sure whether there's already a pattern somewhere in Gardener that would allow doing so in a race-free way?)

**Comment by MichaelEischer** [comment]
(I'll add some documentation next week)

**Comment by MichaelEischer** [comment]
@rfranzke I've fixed the tests and addressed the review comments. PTAL.

File: plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Configuration provides configuration for the ShootResourceReservation admission controller.
+type Configuration struct {
+	metav1.TypeMeta
+	// UseGKEFormula enables the calculation of resource reservations based on
+	// the CPU and memory resources available for a machine type.
+	UseGKEFormula bool `json:"useGKEFormula,omitempty"`
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go]: Why did you add `omitempty` here?

File: plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	metav1.TypeMeta
+	// UseGKEFormula enables the calculation of resource reservations based on
+	// the CPU and memory resources available for a machine type.
+	UseGKEFormula bool `json:"useGKEFormula,omitempty"`
```
Comment (rfranzke) [inline plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/v1alpha1/types.go]: same q here

File: plugin/pkg/shoot/resourcereservation/admission.go (hunk 5 above)
Comment (rfranzke) [reply plugin/pkg/shoot/resourcereservation/admission.go]: Where is this function (I cannot find it 👀)? 

> Or did you have in mind to cache the map? (I'm not sure whether there's already a pattern somewhere in Gardener that would allow doing so in a race-free way?)

No, I did not have in mind to cache it - it can run every time the admission hook is called. Though, it should still be better/faster than running through all machine types again and again for each worker pool.

File: plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go (hunk 7 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go]: No particular reason. The configuration fields in other admission plugins with configuration also seem to use `omitempty`, see for example `plugin/pkg/shoot/tolerationrestriction/apis/shoottolerationrestriction/v1alpha1/types.go`. So, that's essentially copy&paste from there. But I can also drop the `omitempty` if you want.

Comment (rfranzke) [reply plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go]: Typically, `omitempty` is only used for optional fields. If this is an optional field, you should change the type to `*bool` as well and add the `// +optional` tag to the comment. In this case, I think it's not optional, so let's remove the `omitempty` then :)

Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/apis/shootresourcereservation/types.go]: Ok, I've removed the omitempty in both places.

**Comment by rfranzke** [comment]
@MichaelEischer Then only the discussion in https://github.com/gardener/gardener/pull/9449#discussion_r1562266452 is unresolved

File: plugin/pkg/shoot/resourcereservation/admission.go (hunk 5 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/admission.go]: The function is at https://github.com/gardener/gardener/pull/9449/files#diff-e5db5c7cca19cd1aa25f19d3df2c16e7e96a831ffe5ad82f4f87a40eca5cfa8aR233 .

Comment (rfranzke) [reply plugin/pkg/shoot/resourcereservation/admission.go]: Hm, seems GitHub was showing me some old version of this PR last week. Anyways, thanks!

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

File: plugin/pkg/shoot/resourcereservation/admission.go (hunk 5 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/resourcereservation/admission.go]: I had the same problem this morning in this PR and ended up totally confused :confused:  .
